module/VuFind/src/VuFind/I18n/Translator/Loader/ExtendedIni.php

```diff
@@ -37,218 +38,209 @@
-     * @throws InvalidArgumentException
+     * ExtendedIni constructor.
      */
-    public function load($locale, $filename)
+    public function __construct()
     {
-        // Reset the loaded files list:
-        $this->resetLoadedFiles();
-
-        // Load base data:
-        $data = $this->loadLanguageLocale($locale, $filename);
-
-        // Load fallback data, if any:
-        if (!empty($this->fallbackLocales)) {
-            foreach ($this->fallbackLocales as $fallbackLocale) {
-                $newData = $this->loadLanguageLocale($fallbackLocale, $filename);
-                $newData->merge($data);
-                $data = $newData;
-            }
-        }
-
-        return $data;
+        $this->reader = new ExtendedIniReader();
     }
 
     /**
-     * Get the language file name for a language and domain
+     * Sets the list of base directories.
      *
-     * @param string $locale Locale name
-     * @param string $domain Text domain (if any)
+     * @param string[] $dirs {@see $dirs}
      *
-     * @return string
+     * @return void
      */
-    public function getLanguageFilename($locale, $domain)
+    public function setDirs(array $dirs)
     {
-        return empty($domain)
-            ? $locale . '.ini'
-            : $domain . '/' . $locale . '.ini';
+        $this->dirs = $dirs;
     }
 
     /**
-     * Reset the loaded file list.
+     * Sets the map of fallback languages.
+     *
+     * @param string[] $fallbacks {@see $fallbacks}
      *
      * @return void
      */
-    protected function resetLoadedFiles()
+    public function setFallbacks(array $fallbacks)
     {
-        $this->loadedFiles = [];
+        $this->fallbacks = $fallbacks;
     }
 
     /**
-     * Check if a file has already been loaded; mark it loaded if it is not already.
+     * Load method defined by RemoteLoaderInterface.
      *
-     * @param string $filename Name of file to check and mark as loaded.
+     * @param string $locale     Locale
+     * @param string $textDomain Text domain
      *
-     * @return bool True if loaded, false if new.
+     * @return TextDomain
      */
-    protected function checkAndMarkLoadedFile($filename)
+    public function load($locale, $textDomain = 'default')
     {
-        if (isset($this->loadedFiles[$filename])) {
-            return true;
-        }
-        $this->loadedFiles[$filename] = true;
-        return false;
+        $this->dict = $this->list = [];
+        $this->loadLocale($locale, $textDomain, $textDomain === 'default');
+
+        return array_reduce(
+            array_reverse($this->list),
+            function (TextDomain $data, $filename) {
+                return $data->merge($this->dict[$filename]['data']);
+            }, new TextDomain(
+                [
+                    self::INFO => [
+                        'dirs' => $this->dirs,
+                        'list' => $this->list,
+                        'dict' => $this->dict
+                    ]
```

Comment:
Aside from the actual translations the returned text domain will contain all data necessary to retrace the results - which files have been loaded/merged in which order etc.)

This will be handy for debugging (missing, faulty) translations in more complex scenarios extensively using language file inheritance.

Reply:
Is there any benefit to creating a flag for this detailed tracing so it's only generated in debug mode, or would checking conditions to perform a trace actually be more expensive than simply doing it all the time?